src/Temporalio/Worker/WorkflowInstance.cs

```diff
@@ -841,6 +866,145 @@ private void Apply(WorkflowActivationJob job)
 
         private void ApplyCancelWorkflow() => cancellationTokenSource.Cancel();
 
+        private void ApplyDoUpdate(DoUpdate update)
+        {
+            // Queue it up so it can run in workflow environment
+            _ = QueueNewTaskAsync(() =>
+            {
+                HandleUpdateInput updateInput;
+
+                // Get query, convert args, do validation
+                try
+                {
+                    // Try to find the update definition
+                    var updates = mutableUpdates.IsValueCreated ? mutableUpdates.Value : Definition.Updates;
+                    if (!updates.TryGetValue(update.Name, out var updateDefn))
+                    {
+                        updateDefn = DynamicUpdate;
+                        if (updateDefn == null)
+                        {
+                            var knownUpdates = updates.Keys.OrderBy(k => k);
+                            throw new InvalidOperationException(
+                                $"Update handler for {update.Name} expected but not found, " +
+                                $"known updates: [{string.Join(" ", knownUpdates)}]");
+                        }
+                    }
+                    // Build update input
+                    updateInput = new(
+                        Id: update.Id,
+                        Update: update.Name,
+                        Definition: updateDefn,
+                        Args: DecodeArgs(
+                            method: updateDefn.Method ?? updateDefn.Delegate!.Method,
+                            payloads: update.Input,
+                            itemName: $"Update {update.Name}",
+                            dynamic: updateDefn.Dynamic,
+                            dynamicArgPrepend: update.Name),
+                        Headers: update.Headers);
+                    // If validation is requested, perform it inline. We call the interceptor
+                    // regardless of whether a validator is defined. An interceptor can validate
+                    // even if there is no validator method.
+                    if (update.RunValidator)
+                    {
+                        // Capture command count so we can ensure it is not altered during validation
+                        var origCmdCount = completion?.Successful?.Commands?.Count ?? 0;
+                        inbound.Value.ValidateUpdate(updateInput);
```

Comment:
Seems this would be useful to put behind some kind of "read only mode" protector like exists in Python / Java?

Reply:
Yeah, this was only the second time I needed such a thing and the error message is different enough and impl simpl;e enough that I didn't make reuseable (if .NET had simple yield-based contextmanager type thing like Python, definitely).